### ISSUE TITLE
Eviction error for downgrade of Scala 3.x

### ISSUE THREAD

**Issue by eed3si9n** [sbt/sbt#6694]
## steps

```scala
ThisBuild / scalaVersion := "3.0.0"
```

pick some library that uses `3.1.0`.

## problem

sbt would likely pick 3.0.0, which might result in some error.

## expectation

Eviction warning or error?

## notes

https://web.archive.org/web/20211019043957/https://twitter.com/eed3si9n/status/1450320726750375937

<img width="629" height="442" alt="Eugene wrote at some point sbt decided scalaVersion set by the user is absolute, partly due to Typelevel Scala, compiler plugins, and macros etc. I don't think the lack of forward compat would change that, but perhaps the user could get a heads-up via eviction error? quoting a tweet by Seth: PSA: Scala 3.0 cannot consume 3.1-compiled libraries. Scala 3 minor versions are backwards compatible only, not forwards. This is a big change from Scala 2." src="https://github.com/user-attachments/assets/177bd08a-ca6c-457d-af5e-13e1d529a3d8" />

**Comment by SethTisue** [comment]
Dale points out that it might not be necessary, as some of us initially assumed, to look inside the JAR and glean the TASTy version from looking at a class file. sbt could look at the POM and see what scala3-library version it declares a dependency on, yes?

**Comment by romanowski** [comment]
I was thinking to implement a similar thing in coursier. Basically, after dependency is done, check what are the Scala dependencies in all resolved libraries (including transitive ones) and reports error for every library that requires newer Scala. 

We are now exploring alternative solutions within the compiler itself so for me, this eviction-base approach is something that we call to fall back to if other options won't be feasible.

**Comment by dwijnand** [comment]
Looking at the transitive scala3-library version is a reasonable and easier approximation.

I also wonder if it's time to let transitive dependencies update the scala version.  We don't seem to mind when transitive dependencies update our declare dependencies, and we have ways to force that not to happen, so why not this one too?  Why is it better for me to be running 3.0.1 compiler with a dependency that used 3.0.2?

**Comment by mpilquist** [comment]
> I also wonder if it's time to let transitive dependencies update the scala version. We don't seem to mind when transitive dependencies update our declare dependencies, and we have ways to force that not to happen, so why not this one too? Why is it better for me to be running 3.0.1 compiler with a dependency that used 3.0.2?

Strongly in favor of this proposal!

**Comment by eed3si9n** [comment]
Prior to #2634, `scalaVersion` controlled the Scala version used for the `compile` task, which also brings in modules needed by the compiler, and `scala-library` dependency was a soft dependency initially matching the `scalaVersion` but can be evicted. Over time, tooling authors and compiler plugin authors wanted align the scala-library version with scala-compiler, scala-reflect etc versions, so we now enforce scala-library version strongly.

If we let scala-library versions become a soft suggestion again, "like any other dependency" then scala-compiler version would mismatch against scala-library again. I don't know if letting scala-compiler toolchain dependency slip and slide is a good idea either. In the history of sbt, and all other build tools that I know of, the compiler version is something the user specifies, in part because there are specific features and compile-time differences.

I feel like letting the user pick one version of `scalaVersoin` and failing at compile-time or build time saying "hey you need to bump it up to 3.1.0" is actually not a bad situation to be in.

**Comment by dwijnand** [comment]
> If we let scala-library versions become a soft suggestion again, "like any other dependency" then scala-compiler version would mismatch against scala-library again

Oh I meant we keep that alignment: a dependency using scala3-library 3.0.2 would mean we run scala3-compiler 3.0.2.

(But, btw, because the library is backwards compatible it means that old compilers should continue to work with newer standard libraries.)

**Comment by xuwei-k** [comment]
- https://github.com/scala/scala3/issues/25406

**Comment by bitloi** [comment]
@eed3si9n Can I pick the issue up?

**Comment by eed3si9n** [comment]
@bitloi Yes, please. Here's a quick write up of my understanding/thinking. There are essentially two degrees of compatibility when it comes to `scalaVersion` (the compiler toolchain used to compile `*.scala`) and the standard library (`scala3-library_3` for Scala 3.x).

First is _runtime_ binary compatibility. According to [Binary Compatibility of Scala Releases](https://docs.scala-lang.org/overviews/core/binary-compatibility-of-scala-releases.html):
> When two versions of Scala are binary compatible, it is safe to compile your project on one Scala version and link against another Scala version at run time.

1. Scala 3 guarantees both **backward and forward** compatibility across **patch** releases within a single minor release. In particular, this applies within an entire LTS releases such as Scala 3.3.x.
2. Scala 3 also guarantees **backward** compatibility across **minor** releases in the entire 3.x series, but not forward compatibility. This means that libraries compiled with any Scala 3.x version can be used in projects compiled with any Scala 3.y version with `y >= x`.

Second is _compile-time_ alignment. If we phrase it as something like:

1. During the compile-time, Scala compiler should use the exact version of standard libraries on the classpath, which includes `scala3-library_3` for Scala 3.x, and `scala-library` for Scala 2.x and 3.x after 3.8+.

Here's an example table we can use for testing:

| scalaVersion | found scala3-library_3 | binary compatibility | compile-time alignment |
|-------------:|-----------------------:|:-------------:|:-------------:|
|      `3.3.3` |                `3.3.0` |       n/a*1       |       ❌       |
|      `3.3.3` |                `3.3.3` |       ✅       |       ✅       |
|      `3.3.3` |                `3.3.7` |       ✅       |      ❌       |
|      `3.3.3` |                `3.8.1` |       ❌       |       ❌       |
|      `3.8.1` |                `3.3.3` |       n/a*1       |       ❌       |
|      `3.8.1` |                `3.8.2` |       ✅       |       ❌       |
|      `3.8.2` |                `3.8.2` |       ✅       |       ✅        |

*1) This shouldn't happen since sbt adds `scala3_library_3` when auto library is `true`.

We can adopt the error message that's in use for Scala 2.13 (https://github.com/sbt/sbt/pull/7480) and say something like

```
s"""`${name.value}/scalaVersion` needs to be upgraded to $libVer. 
   |To maintain the compile-time alignment, the Scala compiler cannot be older than $libName on the
   |dependency classpath. See `${name.value}/evicted` why $libName was upgraded from $sv to $libVer.
   |""".stripMargin
```

As per usual please check out our [Contributor's Guildeline](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md).
